### ISSUE TITLE
taking second derivatives

### ISSUE THREAD

**Issue by tungli** [tinygrad/tinygrad#2584]
I cannot figure out how to compute the second derivative. For example, after:
```python
from tinygrad.nn import Tensor

x = Tensor(2., requires_grad=True)
y = (3.*x).exp()
y.backward()
dydx = x.grad    # first derivative.. no problem
```

I would like to do something like:
```python
dydx.backward()
d2ydx2 = x.grad
```

However this fails to update the gradient, as `dydx.numpy() == d2ydx2.numpy()`.

In torch this works by using `y.backward(create_graph=True)`.

**Comment by geohot** [comment]
The grads should be normal tensors, though you probably have to mark them as requires grad. Want to add a failing test and I'll look?

**Comment by tungli** [comment]
Thanks! The test is just:
```python
from tinygrad.nn import Tensor
import numpy as np

x = Tensor(2., requires_grad=True)
y = (3.*x).exp()

y.requires_grad = True

y.backward()
dydx = x.grad
dydx.requires_grad = True

#x.grad = None
dydx.backward()
d2ydx2 = x.grad

assert np.isclose(d2ydx2.numpy(), 9*y.numpy())
```

The working version of the above in pytorch looks like this:
```python
from torch import tensor as Tensor
import numpy as np

x = Tensor(2., requires_grad=True)
y = (3.*x).exp()

# use of torch.autograd.grad is recommended instead of this...
y.backward(create_graph=True)
dydx = x.grad

x.grad = None      # otherwise grad is dydx + d2ydx2
dydx.backward()
d2ydx2 = x.grad

assert np.isclose(d2ydx2.detach().numpy(), 9*y.detach().numpy())
```

**Comment by geohot** [comment]
Err, we don't have a create_graph=True, I think you'd have to look into what that does.

This is not high priority for us, but if you wrote a PR with good tests I'd merge it

**Comment by tungli** [comment]
Thanks for checking George!

This comes up in Physics-informed NNs (PINNs). I will have a look if I can implement this but no promises.
